src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedList`1.cs

```diff
@@ -0,0 +1,794 @@
+            }
+
+            if (initialCapacity > 0)
+            {
+                initialCapacity = _segmentSize * ((initialCapacity + _segmentSize - 1) / _segmentSize);
+                _items = new T[initialCapacity >> _segmentShift][];
+                for (var i = 0; i < _items.Length; i++)
+                {
+                    _items[i] = new T[_segmentSize];
+                }
+
+                _capacity = initialCapacity;
+            }
+        }
+
+        /// <summary>
+        /// Returns the count of elements in the list.
+        /// </summary>
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        /// <summary>
+        /// Copy to Array
+        /// </summary>
+        /// <returns>Array copy</returns>
+        public T[] UnderlyingArray => ToArray();
+
+        /// <summary>
+        /// Returns the last element on the list and removes it from it.
+        /// </summary>
+        /// <returns>The last element that was on the list.</returns>
+        public T Pop()
+        {
+            if (_count == 0)
+            {
+                throw new InvalidOperationException("Attempting to remove an element from empty collection.");
+            }
+
+            var oldSegmentIndex = --_count >> _segmentShift;
+            var result = _items[oldSegmentIndex][_count & _offsetMask];
+
+            var newSegmentIndex = (_count - 1) >> _segmentShift;
+
+            if (newSegmentIndex != oldSegmentIndex)
+            {
+                _items[oldSegmentIndex] = null;
+                _capacity -= _segmentSize;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Returns true if this ICollection is read-only.
+        /// </summary>
+        bool ICollection<T>.IsReadOnly
+        {
+            get { return false; }
+        }
+
+        /// <summary>
+        /// Gets or sets the given element in the list.
+        /// </summary>
+        /// <param name="index">Element index.</param>
+        public T this[int index]
+        {
+            get
+            {
+                return _items[index >> _segmentShift][index & _offsetMask];
+            }
+
+            set
+            {
+                _items[index >> _segmentShift][index & _offsetMask] = value;
+            }
+        }
+
+        /// <summary>
+        /// Necessary if the list is being used as an array since it creates the segments lazily.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns>true if the segment is allocated and false otherwise</returns>
+        public bool IsValidIndex(int index)
+        {
+            return _items[index >> _segmentShift] != null;
+        }
+
+        /// <summary>
+        /// Get slot of an element
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="slot"></param>
+        /// <returns></returns>
+        public T[] GetSlot(int index, out int slot)
```

Comment:
would prefer to make thsi as private as possible and not expose internals liek this unless we def have a consumer for it.